File: source/math/Float3.java
package math;

import java.awt.*;
import java.io.Serializable;

public class Float3 implements Serializable {
    public float x;
    public float y;
    public float z;

    public Float3() {
        set(0);
    }

    public Float3(float a) {
        set(a);
    }

    public Float3(float[] data) {
        set(data);
    }

    public Float3(float x, float y, float z) {
        set(x, y, z);
    }

    public Float3(Float2 v, float z) {
        set(v, z);
    }

    public Float3(float x, Float2 v) {
        set(x, v);
    }

    public Float3(Float3 v) {
        set(v);
    }

    public Float3(Int3 v) {
        set(v);
    }

    public Float3(Color color) {
        set(color);
    }

    // Setters
    public void set(float x, float y, float z) {
        this.x = x; this.y = y; this.z = z;
    }

    public void set(float a) {
        set(a, a, a);
    }

    public void set(float[] data) {
        set(data[0], data[1], data[2]);
    }

    public void set(Float2 v, float z) {
        set(v.x, v.y, z);
    }

    public void set(float x, Float2 v) {
        set(x, v.x, v.y);
    }

    public void set(Float3 v) {
        set(v.x, v.y, v.z);
    }

    public void set(Int3 v) {
        set(v.x, v.y, v.z);
    }

    public void set(Color color) {
        final float conv = 1.0f / 255.0f;
        set(color.getRed() * conv, color.getGreen() * conv, color.getBlue() * conv);
    }

    // Getters
    public Float2 xy() {
        return new Float2(x, y);
    }

    public float[] array() {
        return new float[] { x, y, z };
    }

    public Color color() {
        return new Color(
            (int) Math.min(Math.max(x * 255, 0), 255),
            (int) Math.min(Math.max(y * 255, 0), 255),
            (int) Math.min(Math.max(z * 255, 0), 255)
        );
    }

    // Math
    public Float3 add(Float3 v) {
        return new Float3(x + v.x, y + v.y, z + v.z);
    }

    public Float3 subtract(Float3 v) {
        return new Float3(x - v.x, y - v.y, z - v.z);
    }

    public Float3 multiply(float a) {
        return new Float3(x * a, y * a, z * a);
    }

    public Float3 multiply(Float3 v) {
        return new Float3(x * v.x, y * v.y, z * v.z);
    }

    public Float3 divide(float a) {
        return multiply(1.0f / a);
    }

    public Float3 divide(Float3 v) {
        return new Float3(x / v.x, y / v.y, z / v.z);
    }

    public boolean equals(Float3 v) {
        return x == v.x && y == v.y && z == v.z;
    }

    public Float3 negate() {
        return multiply(-1.0f);
    }

    public Float3 absolute() {
        return new Float3(Math.abs(x), Math.abs(y), Math.abs(z));
    }

    public float length() {
        return (float) Math.sqrt(dot(this));
    }

    public Float3 normalize() {
        return divide(length());
    }

    public float dot(Float3 v) {
        return x * v.x + y * v.y + z * v.z;
    }

    public Float3 cross(Float3 v) {
        return new Float3(y * v.z - z * v.y, z * v.x - x * v.z, x * v.y - y * v.x);
    }

    public float angle(Float3 v) {
        return (float) Math.toDegrees(Math.acos(normalize().dot(v.normalize())));
    }

    public Float3 rotate(float angle, Float3 axis) {
        final float angleSin = (float) Math.sin(Math.toRadians(angle * 0.5f));
        final float angleCos = (float) Math.cos(Math.toRadians(angle * 0.5f));

        final float qx = axis.x * angleSin;
        final float qy = axis.y * angleSin;
        final float qz = axis.z * angleSin;
        final float x2 = qx * qx;
        final float y2 = qy * qy;
        final float z2 = qz * qz;
        final float w2 = angleCos * angleCos;
        final float xy = qx * qy;
        final float xz = qx * qz;
        final float yz = qy * qz;
        final float xw = qx * angleCos;
        final float yw = qy * angleCos;
        final float zw = qz * angleCos;

        Float3 temp = new Float3(0);
        temp.x = (w2 + x2 - z2 - y2) * x + (-zw + xy - zw + xy) * y + (yw + xz + xz + yw) * z;
        temp.y = (xy + zw + zw + xy) * x + ( y2 - z2 + w2 - x2) * y + (yz + yz - xw - xw) * z;
        temp.z = (xz - yw + xz - yw) * x + ( yz + yz + xw + xw) * y + (z2 - y2 - x2 + w2) * z;
        return temp;
    }

    public Float3 reflect(Float3 v) {
        final Float3 normal = v.normalize();
        return this.subtract(normal.multiply(2 * this.dot(normal)));
    }

    public static Float3 getPosX() {
        return new Float3(1.0f, 0.0f, 0.0f);
    }
    public static Float3 getNegX() {
        return new Float3(-1.0f, 0.0f, 0.0f);
    }
    public static Float3 getPosY() {
        return new Float3(0.0f, 1.0f, 0.0f);
    }
    public static Float3 getNegY() {
        return new Float3(0.0f, -1.0f, 0.0f);
    }
    public static Float3 getPosZ() {
        return new Float3(0.0f, 0.0f, 1.0f);
    }
    public static Float3 getNegZ() {
        return new Float3(0.0f, 0.0f, -1.0f);
    }

    public String toString() {
        return "(" + x + ", " + y + ", " + z + ")";
    }
}

File: source/math/Float3.kt
package math

import java.awt.Color
import java.io.Serializable
import kotlin.math.acos
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

class Float3 : Serializable {
    var x: Float = 0f
    var y: Float = 0f
    var z: Float = 0f

    constructor(x: Float, y: Float, z: Float) {
        this.x = x
        this.y = y
        this.z = z
    }

    constructor() : this(0f, 0f, 0f)

    constructor(a: Float) : this(a, a, a)

    constructor(data: FloatArray) : this(data[0], data[1], data[2])

    constructor(v: Float2, z: Float) : this(v.x, v.y, z)

    constructor(x: Float, v: Float2) : this(x, v.x, v.y)

    constructor(v: Float3) : this(v.x, v.y, v.z)

    constructor(v: Int3) : this(v.x.toFloat(), v.y.toFloat(), v.z.toFloat())

    constructor(color: Color) {
        val conv = 1f / 255f
        x = color.red * conv
        y = color.green * conv
        z = color.blue * conv
    }

    // Getters
    val length: Float
        get() = sqrt(this * this)

    val xy: Float2
        get() = Float2(x, y)

    val array: FloatArray
        get() = floatArrayOf(x, y, z)

    val color: Color
        get() = Color(
            (x * 255).toInt().coerceAtLeast(0).coerceAtMost(255),
            (y * 255).toInt().coerceAtLeast(0).coerceAtMost(255),
            (z * 255).toInt().coerceAtLeast(0).coerceAtMost(255)
        )

    // Math
    operator fun plus(v: Float3): Float3 {
        return Float3(x + v.x, y + v.y, z + v.z)
    }

    operator fun minus(v: Float3): Float3 {
        return Float3(x - v.x, y - v.y, z - v.z)
    }

    operator fun times(a: Float): Float3 {
        return Float3(x * a, y * a, z * a)
    }

    operator fun times(v: Float3): Float {
        return x * v.x + y * v.y + z * v.z
    }

    infix fun x(v: Float3): Float3 {
        return Float3(y * v.z - z * v.y, z * v.x - x * v.z, x * v.y - y * v.x)
    }

    operator fun div(a: Float): Float3 {
        return this * (1f / a)
    }

    operator fun unaryMinus(): Float3 {
        return this * -1f
    }

    override fun equals(other: Any?): Boolean {
        if (other is Float3) {
            return x == other.x && y == other.y && z == other.z
        }
        return false
    }

    fun angle(v: Float3): Float {
        return toDegrees(acos(normalize(this) * normalize(v)))
    }

    fun rotate(angle: Float, axis: Float3): Float3 {
        val angleSin = sin(toRadians(angle * 0.5f))
        val angleCos = cos(toRadians(angle * 0.5f))
        val qx = axis.x * angleSin
        val qy = axis.y * angleSin
        val qz = axis.z * angleSin
        val x2 = qx * qx
        val y2 = qy * qy
        val z2 = qz * qz
        val w2 = angleCos * angleCos
        val xy = qx * qy
        val xz = qx * qz
        val yz = qy * qz
        val xw = qx * angleCos
        val yw = qy * angleCos
        val zw = qz * angleCos

        val temp = Float3(0f)
        temp.x = (w2 + x2 - z2 - y2) * x + (-zw + xy - zw + xy) * y + (yw + xz + xz + yw) * z
        temp.y = (xy + zw + zw + xy) * x + (y2 - z2 + w2 - x2) * y + (yz + yz - xw - xw) * z
        temp.z = (xz - yw + xz - yw) * x + (yz + yz + xw + xw) * y + (z2 - y2 - x2 + w2) * z
        return temp
    }

    fun reflect(v: Float3): Float3 {
        val normal = normalize(v)
        return this - (normal * (this * normal * 2f))
    }

    override fun toString(): String {
        return "($x, $y, $z)"
    }

    override fun hashCode(): Int {
        var result = x.hashCode()
        result = 31 * result + y.hashCode()
        result = 31 * result + z.hashCode()
        return result
    }

    companion object {
        val posX: Float3
            get() = Float3(1f, 0f, 0f)
        val negX: Float3
            get() = Float3(-1f, 0f, 0f)
        val posY: Float3
            get() = Float3(0f, 1f, 0f)
        val negY: Float3
            get() = Float3(0f, -1f, 0f)
        val posZ: Float3
            get() = Float3(0f, 0f, 1f)
        val negZ: Float3
            get() = Float3(0f, 0f, -1f)
    }
}